File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;
import java.io.*;

class Solution {
    boolean solution(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        char[] a = s.toCharArray();
    
        if(a[0] == ')') 
            return false;
         
        for(int i = 0; i < a.length; i++){
            if(stack.isEmpty()){
                stack.push(a[i]);
                continue;
            }
            
            char target = stack.pop();
            // if (target == a[i]) {
            //     stack.push(target);
            //     stack.push(a[i]);
            // }
            if (!(target == '(' && a[i] == ')')) {
                stack.push(target);
                stack.push(a[i]);
            }
            
        }
        
        boolean answer = (stack.isEmpty()) ? true : false ;

        return answer;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
function solution(s){
    let stack = [];
    
    if(s.startsWith(')'))
        return false;
    
    for(let el of s){
        if(stack.length == 0){
            stack.push(el);
            console.log(stack);
            continue;
        }
        
        let pop = stack.pop();
        if(pop === '(' && el === '(' || pop === ')' && el === ')' || pop.startsWith(')')){
            stack.push(pop);
            stack.push(el);
        } 
    }
    
    return stack.length == 0 ? true : false;
}